File: orchestrator/app/api/schemas/archive.py
"""Archive API DTOs."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

from orchestrator.app.api.schemas.common import ApiMeta, EmptyState, Pagination


class ArchiveItemResponse(BaseModel):
    ad_id: str
    job_id: str | None = None
    title: str
    thumbnail_url: str | None = None
    image_url: str | None = None
    status: Literal["generating", "saved", "favorite", "failed"] = "saved"
    ad_format: str | None = None
    platform: str | None = None
    source: str = "generated"
    created_at: str | None = None
    saved_at: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class ArchiveItemCreateRequest(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    public_job_id: str | None = Field(default=None, max_length=120)
    thumbnail_url: str | None = None
    image_url: str | None = None
    status: Literal["saved", "favorite", "failed"] = "saved"
    ad_format: str | None = Field(default=None, max_length=80)
    platform: str | None = Field(default=None, max_length=80)
    source: Literal["generated", "reference_template", "uploaded"] = "generated"
    workspace_id: str | None = None
    user_id: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class ArchiveListResponse(BaseModel):
    success: Literal[True] = True
    items: list[ArchiveItemResponse]
    pagination: Pagination
    empty_state: EmptyState | None = None
    meta: ApiMeta = Field(default_factory=ApiMeta)


class ArchiveMutationResponse(BaseModel):
    success: Literal[True] = True
    item: ArchiveItemResponse
    meta: ApiMeta = Field(default_factory=ApiMeta)

File: apps/bff/src/schemas/archive.js
import { z } from "zod";

export const archiveItemSchema = z.object({
  title: z.string().trim().min(1), publicJobId: z.string().trim().min(1).optional(), public_job_id: z.string().trim().min(1).optional(),
  imageUrl: z.string().trim().min(1).optional().nullable(), image_url: z.string().trim().min(1).optional().nullable(),
  thumbnailUrl: z.string().trim().min(1).optional().nullable(), thumbnail_url: z.string().trim().min(1).optional().nullable(),
  status: z.enum(["saved", "favorite", "failed"]).optional(), adFormat: z.string().trim().min(1).optional().nullable(),
  ad_format: z.string().trim().min(1).optional().nullable(), platform: z.string().trim().min(1).optional().nullable(),
  source: z.enum(["generated", "reference_template", "uploaded"]).optional(), workspaceId: z.string().trim().min(1).optional(),
  workspace_id: z.string().trim().min(1).optional(), userId: z.string().trim().min(1).optional(), user_id: z.string().trim().min(1).optional(),
  metadata: z.record(z.unknown()).optional()
});
export const archiveItemUpdateSchema = z.object({ status: z.enum(["saved", "favorite"]) });

const compactObject = (value) => Object.fromEntries(Object.entries(value).filter(([, item]) => item !== undefined && item !== null));
export function toArchiveItemPayload(data) {
  return compactObject({
    title: data.title, public_job_id: data.public_job_id ?? data.publicJobId, image_url: data.image_url ?? data.imageUrl,
    thumbnail_url: data.thumbnail_url ?? data.thumbnailUrl, status: data.status ?? "saved", ad_format: data.ad_format ?? data.adFormat,
    platform: data.platform, source: data.source, workspace_id: data.workspace_id ?? data.workspaceId,
    user_id: data.user_id ?? data.userId, metadata: data.metadata
  });
}
